idseq_dag/util/dict.py

```diff
@@ -17,65 +19,115 @@ class IdSeqDict(object):
         a key -> value permanent dictionary store with sqlite3 as the backend.
         key has to be a string
         value could be either array of strings or a string
+        this version is read-only;  to write, use IdSeqDictWritable
+        use only as a WITH context, which is equivalent to the context of the shared db connection,
+        protected by a multiprocessing lock
     '''
-    def __init__(self, db_path, value_type, read_only=True):
+    def __init__(self, db_path, value_type):
         self.db_path = db_path
         self.value_type = value_type
-        self.read_only = read_only
-        self.uri_db_path = pathlib.Path(self.db_path).as_uri()
-        if self.read_only:
-            self.uri_db_path += "?mode=ro&immutable=1&nolock=1&cache=private"
-        with log.log_context(f"db_open", {"db_path": self.db_path, "read_only": self.read_only, "uri_db_path": self.uri_db_path}):
-            self.db_conn = sqlite3.connect(self.uri_db_path, check_same_thread=False, uri=True) # make it thread safe
-        self._open = True
-        with log.log_context(f"db_assert_table", {"db_path": self.db_path, "read_only": self.read_only}):
-            cursor = self.db_conn.cursor()
-            if not self.read_only:
-                res = cursor.execute(f"SELECT count(*) FROM sqlite_master WHERE type='table' AND name='{SQLITE_TABLE_NAME}';")
+        self.db_conn = None
+        self.db_entered = False
+        self.lock = multiprocessing.RLock()
+
+    def get_uri(self):
+        ''' In this class, URI for read-only access.  In the writable class, URI for read and write access. '''
+        return pathlib.Path(self.db_path).as_uri() + "?mode=ro&immutable=1&nolock=1&cache=private"
+
+    def connect():
+        assert self.db_conn == None
+        uri_db_path = self.get_uri()
+        with log.log_context(f"db_open", {"db_path": self.db_path, "uri_db_path": uri_db_path}):
+            self.db_conn = sqlite3.connect(uri_db_path, uri=True)
+
+    def ensure_table_exists():
```

Comment:
In the read-only class, this ensures early failure if the table is missing.   In the IDSeqDictForUpdate derived class, this method is overridden to create the table if necessary.